File: app/src/main/java/com/fitbudget/app/domain/MealPlanGenerator.kt
package com.fitbudget.app.domain

import com.fitbudget.app.data.database.entity.FoodEntity
import com.fitbudget.app.data.database.entity.MealEntryEntity
import com.fitbudget.app.domain.model.DietPreference
import com.fitbudget.app.domain.model.FoodRole
import com.fitbudget.app.domain.model.MealType

/**
 * Builds one day's meal plan from the offline food database.
 *
 * Priorities, in order:
 *  1. respect the user's diet preference and food exclusions,
 *  2. stay inside the daily rupee budget,
 *  3. get as close as possible to the estimated calorie target,
 *  4. prefer the cheapest protein per rupee.
 *
 * The generator is a pure function and deterministic for a given day, so re-opening the app
 * never reshuffles today's plan, while different days still get some variety.
 */
object MealPlanGenerator {

    data class Request(
        val epochDay: Long,
        val calorieTarget: Double,
        val dailyBudget: Double,
        val preference: DietPreference,
        val excludedKeys: Set<String> = emptySet(),
        val mealTimes: Map<MealType, Int> = emptyMap()
    )

    /** Share of the daily budget each meal may use. */
    private val budgetShare = mapOf(
        MealType.BREAKFAST to 0.26,
        MealType.LUNCH to 0.32,
        MealType.SNACK to 0.12,
        MealType.DINNER to 0.30
    )

    /** Roles each meal is built from, in plate order. */
    private val mealComposition: Map<MealType, List<FoodRole>> = mapOf(
        MealType.BREAKFAST to listOf(FoodRole.STAPLE, FoodRole.PROTEIN, FoodRole.DAIRY),
        MealType.LUNCH to listOf(FoodRole.STAPLE, FoodRole.PROTEIN, FoodRole.VEGETABLE),
        MealType.SNACK to listOf(FoodRole.SNACK, FoodRole.FRUIT),
        MealType.DINNER to listOf(FoodRole.STAPLE, FoodRole.PROTEIN, FoodRole.VEGETABLE)
    )

    /** Items in these roles can be dropped first when the day does not fit the budget. */
    private val optionalRoles = setOf(FoodRole.FRUIT, FoodRole.DAIRY, FoodRole.BEVERAGE, FoodRole.SNACK)

    private const val MAX_STAPLE_SERVINGS = 5.0

    fun generate(request: Request, foods: List<FoodEntity>): List<MealEntryEntity> {
        val available = foods.filter { food ->
            food.category.allowedFor(request.preference) &&
                food.nameKey !in request.excludedKeys
        }
        if (available.isEmpty()) return emptyList()

        val byRole = available.groupBy { it.role }
        val draft = mutableListOf<DraftItem>()

        for (mealType in MealType.entries) {
            val mealCalories = request.calorieTarget * mealType.calorieShare
            val mealBudget = request.dailyBudget * (budgetShare[mealType] ?: 0.25)
            draft += buildMeal(
                mealType = mealType,
                roles = mealComposition[mealType] ?: listOf(FoodRole.STAPLE, FoodRole.PROTEIN),
                byRole = byRole,
                fallback = available,
                calorieTarget = mealCalories,
                budget = mealBudget,
                seed = request.epochDay + mealType.ordinal * 7L
            )
        }

        trimToBudget(draft, request.dailyBudget)

        return draft.mapIndexed { index, item ->
            val time = request.mealTimes[item.mealType]
                ?: (item.mealType.defaultHour * 60 + item.mealType.defaultMinute)
            MealEntryEntity(
                epochDay = request.epochDay,
                mealType = item.mealType,
                mealLabel = item.mealType.label,
                timeMinutes = time,
                foodId = item.food.id,
                foodName = item.food.name,
                servingLabel = item.food.servingLabel,
                quantity = item.quantity,
                caloriesPerServing = item.food.calories,
                proteinPerServing = item.food.proteinG,
                costPerServing = item.food.costRupees,
                completed = false,
                sortOrder = index
            )
        }
    }

    private data class DraftItem(
        val mealType: MealType,
        val food: FoodEntity,
        var quantity: Double
    ) {
        val calories: Double get() = food.calories * quantity
        val cost: Double get() = food.costRupees * quantity
    }

    private fun buildMeal(
        mealType: MealType,
        roles: List<FoodRole>,
        byRole: Map<FoodRole, List<FoodEntity>>,
        fallback: List<FoodEntity>,
        calorieTarget: Double,
        budget: Double,
        seed: Long
    ): List<DraftItem> {
        val items = mutableListOf<DraftItem>()
        val used = mutableSetOf<Long>()

        roles.forEachIndexed { slot, role ->
            val candidates = (byRole[role] ?: emptyList())
                .filter { it.id !in used }
                .ifEmpty { fallback.filter { it.id !in used } }
            if (candidates.isEmpty()) return@forEachIndexed

            val ordered = when (role) {
                FoodRole.PROTEIN, FoodRole.DAIRY -> candidates.sortedByDescending { it.proteinPerRupee }
                else -> candidates.sortedBy { it.costRupees }
            }
            // Rotate through the cheapest/highest-value half so days differ but stay affordable.
            val pool = ordered.take(maxOf(3, ordered.size / 2))
            val picked = pool[((seed + slot * 13L).mod(pool.size.toLong())).toInt()]
            used += picked.id

            val startQuantity = if (role == FoodRole.STAPLE && mealType != MealType.BREAKFAST) 2.0 else 1.0
            items += DraftItem(mealType, picked, startQuantity)
        }

        if (items.isEmpty()) return items

        balance(items, calorieTarget, budget)
        return items
    }

    /** Nudges the staple (or largest) item so the meal lands near its calorie and cost slot. */
    private fun balance(items: MutableList<DraftItem>, calorieTarget: Double, budget: Double) {
        val flexible = items.firstOrNull { it.food.role == FoodRole.STAPLE } ?: items.maxByOrNull { it.calories }
        if (flexible == null || flexible.food.calories <= 0.0) return

        var guard = 0
        while (guard++ < 20) {
            val calories = items.sumOf { it.calories }
            val cost = items.sumOf { it.cost }
            when {
                calories < calorieTarget * 0.9 &&
                    flexible.quantity + 0.5 <= MAX_STAPLE_SERVINGS &&
                    cost + flexible.food.costRupees * 0.5 <= budget * 1.1 ->
                    flexible.quantity += 0.5

                calories > calorieTarget * 1.15 && flexible.quantity - 0.5 >= 0.5 ->
                    flexible.quantity -= 0.5

                cost > budget * 1.15 && flexible.quantity - 0.5 >= 0.5 ->
                    flexible.quantity -= 0.5

                else -> return
            }
        }
    }

    /**
     * Last pass over the whole day: drop optional extras, then shave staple servings, until the
     * plan fits the daily budget. Never removes the last item of a meal.
     */
    private fun trimToBudget(draft: MutableList<DraftItem>, dailyBudget: Double) {
        if (dailyBudget <= 0.0) return

        fun total() = draft.sumOf { it.cost }

        var guard = 0
        while (total() > dailyBudget && guard++ < 40) {
            val removable = draft
                .filter { it.food.role in optionalRoles }
                .filter { candidate -> draft.count { it.mealType == candidate.mealType } > 1 }
                .maxByOrNull { it.cost }
            if (removable != null) {
                draft.remove(removable)
                continue
            }
            val shrinkable = draft
                .filter { it.quantity > 0.5 }
                .maxByOrNull { it.food.costRupees }
                ?: break
            shrinkable.quantity -= 0.5
        }
    }
}

File: ios/Sources/FitBudgetCore/Domain/MealPlanGenerator.swift
import Foundation

/// Builds one day's meal plan from the offline food database.
///
/// Priorities, in order:
///  1. respect the user's diet preference and food exclusions,
///  2. stay inside the daily rupee budget,
///  3. get as close as possible to the estimated calorie target,
///  4. prefer the cheapest protein per rupee.
///
/// Pure and deterministic for a given day, so re-opening the app never reshuffles today's plan
/// while different days still get some variety.
public enum MealPlanGenerator {

    public struct Request: Equatable, Sendable {
        public let day: DayKey
        public let calorieTarget: Double
        public let dailyBudget: Double
        public let preference: DietPreference
        public let excludedKeys: Set<String>
        public let mealTimes: [MealType: Int]

        public init(
            day: DayKey,
            calorieTarget: Double,
            dailyBudget: Double,
            preference: DietPreference,
            excludedKeys: Set<String> = [],
            mealTimes: [MealType: Int] = [:]
        ) {
            self.day = day
            self.calorieTarget = calorieTarget
            self.dailyBudget = dailyBudget
            self.preference = preference
            self.excludedKeys = excludedKeys
            self.mealTimes = mealTimes
        }
    }

    /// Roles each meal is built from, in plate order.
    private static let composition: [MealType: [FoodRole]] = [
        .breakfast: [.staple, .protein, .dairy],
        .lunch: [.staple, .protein, .vegetable],
        .snack: [.snack, .fruit],
        .dinner: [.staple, .protein, .vegetable]
    ]

    /// Items in these roles are dropped first when the day does not fit the budget.
    private static let optionalRoles: Set<FoodRole> = [.fruit, .dairy, .beverage, .snack]

    private static let maxStapleServings: Double = 5

    public static func generate(request: Request, foods: [FoodItem]) -> [PlannedMealItem] {
        let available = foods.filter { food in
            food.category.allowed(for: request.preference)
                && !request.excludedKeys.contains(food.nameKey)
        }
        guard !available.isEmpty else { return [] }

        var byRole: [FoodRole: [FoodItem]] = [:]
        for food in available {
            byRole[food.role, default: []].append(food)
        }

        var draft: [DraftItem] = []
        for mealType in MealType.allCases {
            let mealCalories = request.calorieTarget * mealType.calorieShare
            let mealBudget = request.dailyBudget * mealType.budgetShare
            draft += buildMeal(
                mealType: mealType,
                roles: composition[mealType] ?? [.staple, .protein],
                byRole: byRole,
                fallback: available,
                calorieTarget: mealCalories,
                budget: mealBudget,
                seed: request.day.value + mealType.sortIndex * 7
            )
        }

        trimToBudget(&draft, dailyBudget: request.dailyBudget)

        return draft.enumerated().map { index, item in
            let minutes = request.mealTimes[item.mealType] ?? item.mealType.defaultMinutesOfDay
            return PlannedMealItem(
                day: request.day,
                mealType: item.mealType,
                mealLabel: item.mealType.label,
                minutesOfDay: minutes,
                foodNameKey: item.food.nameKey,
                foodName: item.food.name,
                servingLabel: item.food.servingLabel,
                quantity: item.quantity,
                caloriesPerServing: item.food.calories,
                proteinPerServing: item.food.proteinG,
                costPerServing: item.food.costRupees,
                sortOrder: index
            )
        }
    }

    private struct DraftItem {
        let mealType: MealType
        let food: FoodItem
        var quantity: Double

        var calories: Double { food.calories * quantity }
        var cost: Double { food.costRupees * quantity }
    }

    private static func buildMeal(
        mealType: MealType,
        roles: [FoodRole],
        byRole: [FoodRole: [FoodItem]],
        fallback: [FoodItem],
        calorieTarget: Double,
        budget: Double,
        seed: Int
    ) -> [DraftItem] {
        var items: [DraftItem] = []
        var used: Set<String> = []

        for (slot, role) in roles.enumerated() {
            var candidates = (byRole[role] ?? []).filter { !used.contains($0.nameKey) }
            if candidates.isEmpty {
                candidates = fallback.filter { !used.contains($0.nameKey) }
            }
            guard !candidates.isEmpty else { continue }

            let ordered: [FoodItem]
            switch role {
            case .protein, .dairy:
                // Cheapest protein first.
                ordered = candidates.sorted { lhs, rhs in
                    lhs.proteinPerRupee == rhs.proteinPerRupee
                        ? lhs.nameKey < rhs.nameKey
                        : lhs.proteinPerRupee > rhs.proteinPerRupee
                }
            default:
                ordered = candidates.sorted { lhs, rhs in
                    lhs.costRupees == rhs.costRupees
                        ? lhs.nameKey < rhs.nameKey
                        : lhs.costRupees < rhs.costRupees
                }
            }

            // Rotate through the best half so days differ but stay affordable.
            let poolSize = max(3, ordered.count / 2)
            let pool = Array(ordered.prefix(poolSize))
            let index = modulo(seed + slot * 13, pool.count)
            let picked = pool[index]
            used.insert(picked.nameKey)

            let startQuantity: Double = (role == .staple && mealType != .breakfast) ? 2 : 1
            items.append(DraftItem(mealType: mealType, food: picked, quantity: startQuantity))
        }

        guard !items.isEmpty else { return items }
        balance(&items, calorieTarget: calorieTarget, budget: budget)
        return items
    }

    /// Nudges the staple (or largest) item so the meal lands near its calorie and cost slot.
    private static func balance(_ items: inout [DraftItem], calorieTarget: Double, budget: Double) {
        let flexibleIndex = items.firstIndex { $0.food.role == .staple }
            ?? items.indices.max(by: { items[$0].calories < items[$1].calories })
        guard let flexibleIndex, items[flexibleIndex].food.calories > 0 else { return }

        var guardCount = 0
        while guardCount < 20 {
            guardCount += 1
            let calories = items.reduce(0) { $0 + $1.calories }
            let cost = items.reduce(0) { $0 + $1.cost }
            let stapleCost = items[flexibleIndex].food.costRupees

            if calories < calorieTarget * 0.9,
               items[flexibleIndex].quantity + 0.5 <= maxStapleServings,
               cost + stapleCost * 0.5 <= budget * 1.1 {
                items[flexibleIndex].quantity += 0.5
            } else if calories > calorieTarget * 1.15, items[flexibleIndex].quantity - 0.5 >= 0.5 {
                items[flexibleIndex].quantity -= 0.5
            } else if cost > budget * 1.15, items[flexibleIndex].quantity - 0.5 >= 0.5 {
                items[flexibleIndex].quantity -= 0.5
            } else {
                return
            }
        }
    }

    /// Last pass over the whole day: drop optional extras, then shave staple servings, until the
    /// plan fits the daily budget. Never removes the last item of a meal.
    private static func trimToBudget(_ draft: inout [DraftItem], dailyBudget: Double) {
        guard dailyBudget > 0 else { return }

        var guardCount = 0
        while draft.reduce(0, { $0 + $1.cost }) > dailyBudget && guardCount < 40 {
            guardCount += 1

            let removable = draft.indices
                .filter { optionalRoles.contains(draft[$0].food.role) }
                .filter { index in
                    draft.filter { $0.mealType == draft[index].mealType }.count > 1
                }
                .max(by: { draft[$0].cost < draft[$1].cost })

            if let removable {
                draft.remove(at: removable)
                continue
            }

            let shrinkable = draft.indices
                .filter { draft[$0].quantity > 0.5 }
                .max(by: { draft[$0].food.costRupees < draft[$1].food.costRupees })

            guard let shrinkable else { break }
            draft[shrinkable].quantity -= 0.5
        }
    }

    /// Always-positive modulo, so a negative seed cannot crash the picker.
    private static func modulo(_ value: Int, _ count: Int) -> Int {
        guard count > 0 else { return 0 }
        let remainder = value % count
        return remainder < 0 ? remainder + count : remainder
    }
}
